File: getImportance.py
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""
#Time Complexity:O(v+e)
#Space Comlexity:O(n)

#DFS Solution:

#Algorithm:
# 1. Create a hashmap to store the employees and their objects.
# 2. write a helper function to add the importance of the employee and their subordinates by calling them recursively. 
# 3. return the result.
class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        if not employees: return 0
        hashMap={e.id:e for e in employees}
        def helper(id):
            result=hashMap[id].importance
            for sub in hashMap[id].subordinates:
                result+=helper(sub)
            return result
        return helper(id)
#         if not employees: return 0
#         hashMap={e.id:e for e in employees}
#         def helper(id):
#             result=hashMap[id].importance
#             for sub in hashMap[id].subordinates:
#                 result+=helper(sub)
#             return result
#         return helper(id)
        

File: getImportance.java
// Time Complexity :O(N)
// Space Complexity :Recursion stack + O(N)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :


/*
// Definition for Employee.
class Employee {
    public int id;
    public int importance;
    public List<Integer> subordinates;
};
*/

class Solution {
    public int getImportance(List<Employee> employees, int id) {
        HashMap<Integer, Employee> map = new HashMap();
        
        for(Employee employee: employees){
            map.put(employee.id , employee);
        }
        
        return helper(map,id);
    }
    
    private int helper(HashMap<Integer, Employee> map , int rootId){
        Employee root = map.get(rootId);
        int total = root.importance;
        
        for(int subordinate : root.subordinates){
            total += helper(map,subordinate);
        }
        return total;
    }
}
